File: torchard/tui/views/edit_branch.py
"""Edit a session's base branch."""

from __future__ import annotations

from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Vertical
from textual.screen import Screen
from textual.widgets import Footer, Input, Label, ListItem, ListView, Static

from torchard.core.db import get_repos
from torchard.core.git import GitError, list_branches
from torchard.core.manager import Manager
from torchard.tui.utils import safe_id


class EditBranchScreen(Screen):
    """Pick a new base branch for a managed session."""

    BINDINGS = [
        Binding("escape", "cancel", "Cancel"),
    ]

    DEFAULT_CSS = """
    EditBranchScreen {
        align: center middle;
    }
    #editbranch-container {
        width: 90;
        height: auto;
        max-height: 50;
        border: solid $accent;
        padding: 1 2;
    }
    #editbranch-title {
        text-align: center;
        color: $accent;
        margin-bottom: 1;
        text-style: bold;
    }
    #editbranch-hint {
        color: $text-muted;
        margin-bottom: 1;
        text-align: center;
    }
    #editbranch-filter {
        margin-bottom: 1;
    }
    #editbranch-list {
        height: 18;
        border: solid $panel;
    }
    #editbranch-error {
        color: $error;
        margin-top: 1;
        text-align: center;
    }
    """

    def __init__(self, manager: Manager, session_id: int, session_name: str) -> None:
        super().__init__()
        self._manager = manager
        self._session_id = session_id
        self._session_name = session_name
        self._branches: list[str] = []
        self._render_seq = 0
        self._id_to_branch: dict[str, str] = {}

    def compose(self) -> ComposeResult:
        with Vertical(id="editbranch-container"):
            yield Static("", id="editbranch-title")
            yield Static("", id="editbranch-hint")
            yield Input(placeholder="Filter or type a branch…", id="editbranch-filter")
            yield ListView(id="editbranch-list")
            yield Static("", id="editbranch-error")
        yield Footer()

    def on_mount(self) -> None:
        self.query_one("#editbranch-title", Static).update(
            f"Edit Branch — [dim]{self._session_name}[/dim]"
        )
        self.query_one("#editbranch-hint", Static).update(
            "Pick the branch new worktrees will branch from."
        )

        # Find the repo for this session
        from torchard.core.db import get_sessions
        sessions = get_sessions(self._manager._conn)
        session = next((s for s in sessions if s.id == self._session_id), None)
        if session is None:
            self.query_one("#editbranch-error", Static).update("[red]Session not found[/red]")
            return

        repos = {r.id: r for r in get_repos(self._manager._conn)}
        repo = repos.get(session.repo_id)
        if repo is None:
            self.query_one("#editbranch-error", Static).update("[red]Repo not found[/red]")
            return

        try:
            self._branches = list_branches(repo.path)
        except GitError as exc:
            self._branches = []
            self.query_one("#editbranch-error", Static).update(str(exc))

        self._populate(self._branches, "")
        self.query_one("#editbranch-filter", Input).focus()

    def _populate(self, branches: list[str], query: str) -> None:
        self._render_seq += 1
        seq = self._render_seq
        self._id_to_branch.clear()
        lv = self.query_one("#editbranch-list", ListView)
        lv.clear()
        for branch in branches:
            widget_id = f"branch-{safe_id(branch)}-{seq}"
            self._id_to_branch[widget_id] = branch
            lv.append(ListItem(Label(branch), id=widget_id))
        if query and query not in branches:
            lv.append(ListItem(
                Label(f"[green]+ Use: [bold]{query}[/bold][/green]"),
                id=f"new-branch-{seq}",
            ))

    def on_input_changed(self, event: Input.Changed) -> None:
        if event.input.id != "editbranch-filter":
            return
        query = event.value.lower()
        filtered = [b for b in self._branches if query in b.lower()]
        self._populate(filtered, event.value)

    def on_input_submitted(self, event: Input.Submitted) -> None:
        if event.input.id != "editbranch-filter":
            return
        lv = self.query_one("#editbranch-list", ListView)
        highlighted = lv.highlighted_child
        if highlighted is None and len(lv.children) > 0:
            highlighted = lv.children[0]
        if highlighted is not None:
            self._select_item(highlighted.id)
        elif event.value.strip():
            self._apply(event.value.strip())

    def on_list_view_selected(self, event: ListView.Selected) -> None:
        self._select_item(event.item.id)

    def _select_item(self, item_id: str | None) -> None:
        if item_id and item_id.startswith("new-branch"):
            typed = self.query_one("#editbranch-filter", Input).value.strip()
            if typed:
                self._apply(typed)
            return
        if item_id and item_id in self._id_to_branch:
            self._apply(self._id_to_branch[item_id])

    def _apply(self, branch: str) -> None:
        try:
            self._manager.set_base_branch(self._session_id, branch)
        except ValueError as exc:
            self.query_one("#editbranch-error", Static).update(f"[red]{exc}[/red]")
            return
        self.app.pop_screen()

    def action_cancel(self) -> None:
        self.app.pop_screen()

File: src/tui/edit_branch.rs
// trellis/src/tui/edit_branch.rs

use crossterm::event::{Event, KeyCode, KeyEvent, KeyEventKind, KeyModifiers};
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders, List, ListItem, ListState, Paragraph};

use crate::git;
use crate::manager::Manager;
use super::{ScreenAction, ScreenBehavior};
use super::theme;

pub struct EditBranchScreen {
    session_id: i64,
    session_name: String,
    // All branches loaded on mount
    all_branches: Vec<String>,
    // Currently displayed (filtered) branches + optional "new branch" entry
    display: Vec<DisplayItem>,
    list_state: ListState,
    filter: String,
    filter_cursor: usize,
    error: String,
}

enum DisplayItem {
    Branch(String),
    NewBranch(String),
}

impl DisplayItem {
    fn label(&self) -> String {
        match self {
            DisplayItem::Branch(b) => b.clone(),
            DisplayItem::NewBranch(b) => format!("+ Use: {}", b),
        }
    }
}

impl EditBranchScreen {
    pub fn new(manager: &Manager, session_id: i64, session_name: String) -> Self {
        let mut screen = Self {
            session_id,
            session_name,
            all_branches: Vec::new(),
            display: Vec::new(),
            list_state: ListState::default(),
            filter: String::new(),
            filter_cursor: 0,
            error: String::new(),
        };
        screen.load_branches(manager);
        screen
    }

    fn load_branches(&mut self, manager: &Manager) {
        // Find the session's repo
        let sessions = manager.get_sessions();
        let session = match sessions.iter().find(|s| s.id == Some(self.session_id)) {
            Some(s) => s.clone(),
            None => {
                self.error = "Session not found".to_string();
                return;
            }
        };

        let repos = manager.get_repos();
        let repo = match repos.iter().find(|r| r.id == Some(session.repo_id)) {
            Some(r) => r.clone(),
            None => {
                self.error = "Repo not found".to_string();
                return;
            }
        };

        match git::list_branches(&repo.path) {
            Ok(branches) => {
                self.all_branches = branches;
            }
            Err(e) => {
                self.error = e.to_string();
            }
        }

        self.rebuild_display();
    }

    fn rebuild_display(&mut self) {
        let query = self.filter.to_lowercase();
        let filtered: Vec<String> = if query.is_empty() {
            self.all_branches.clone()
        } else {
            self.all_branches
                .iter()
                .filter(|b| b.to_lowercase().contains(&query))
                .cloned()
                .collect()
        };

        self.display = filtered
            .into_iter()
            .map(DisplayItem::Branch)
            .collect();

        // Add "New branch" option if filter text is non-empty and not an exact match
        let typed = self.filter.trim().to_string();
        if !typed.is_empty() {
            let is_exact = self.all_branches.iter().any(|b| b == &typed);
            if !is_exact {
                self.display.push(DisplayItem::NewBranch(typed));
            }
        }

        // Reset selection
        if self.display.is_empty() {
            self.list_state.select(None);
        } else {
            let sel = self.list_state.selected().unwrap_or(0).min(self.display.len() - 1);
            self.list_state.select(Some(sel));
        }
    }

    fn apply_selected(&mut self, manager: &mut Manager) -> ScreenAction {
        let idx = match self.list_state.selected() {
            Some(i) => i,
            None => return ScreenAction::None,
        };
        let branch = match self.display.get(idx) {
            Some(DisplayItem::Branch(b)) => b.clone(),
            Some(DisplayItem::NewBranch(b)) => b.clone(),
            None => return ScreenAction::None,
        };
        self.apply_branch(&branch, manager)
    }

    fn apply_branch(&mut self, branch: &str, manager: &mut Manager) -> ScreenAction {
        match manager.set_base_branch(self.session_id, branch) {
            Ok(_) => ScreenAction::Pop,
            Err(e) => {
                self.error = e;
                ScreenAction::None
            }
        }
    }

    fn handle_filter_key(&mut self, code: KeyCode, modifiers: KeyModifiers) {
        super::rename::input_handle_key(&mut self.filter, &mut self.filter_cursor, code, modifiers);
        self.rebuild_display();
    }
}

impl ScreenBehavior for EditBranchScreen {
    fn render(&self, f: &mut Frame, area: Rect, _manager: &Manager) {
        let width = 92u16;
        let list_height = 18u16;
        let height = (list_height + 8).min(area.height.saturating_sub(4));
        let title = format!("Edit Branch — {}", self.session_name);
        let inner = super::rename::render_modal_box(f, area, &title, width, height);

        let chunks = Layout::vertical([
            Constraint::Length(1), // hint
            Constraint::Length(1), // gap
            Constraint::Length(1), // filter label + input
            Constraint::Length(list_height), // list
            Constraint::Length(1), // error
        ])
        .split(inner);

        // Hint
        let hint = Paragraph::new(Span::styled(
            "Pick the branch new worktrees will branch from.",
            Style::default().fg(theme::TEXT_DIM),
        ))
        .alignment(Alignment::Center);
        f.render_widget(hint, chunks[0]);

        // Filter input
        let filter_label = Rect::new(chunks[2].x, chunks[2].y, 8, 1);
        let filter_input_area = Rect::new(chunks[2].x + 8, chunks[2].y, chunks[2].width.saturating_sub(8), 1);
        f.render_widget(
            Paragraph::new(Span::styled("Filter: ", Style::default().fg(theme::TEXT_DIM))),
            filter_label,
        );
        super::rename::render_input(f, filter_input_area, &self.filter, self.filter_cursor);

        // Branch list
        let list_block = Block::default()
            .borders(Borders::ALL)
            .border_style(Style::default().fg(theme::TEXT_DIM));
        let list_inner = list_block.inner(chunks[3]);
        f.render_widget(list_block, chunks[3]);

        let items: Vec<ListItem> = self
            .display
            .iter()
            .map(|item| {
                let style = match item {
                    DisplayItem::Branch(_) => Style::default().fg(theme::TEXT),
                    DisplayItem::NewBranch(_) => Style::default().fg(theme::GREEN),
                };
                ListItem::new(Span::styled(item.label(), style))
            })
            .collect();

        let list = List::new(items)
            .highlight_style(theme::style_cursor())
            .highlight_symbol("> ");

        f.render_stateful_widget(list, list_inner, &mut self.list_state.clone());

        // Error
        if !self.error.is_empty() {
            let err = Paragraph::new(Span::styled(self.error.clone(), theme::style_error()))
                .alignment(Alignment::Center);
            f.render_widget(err, chunks[4]);
        }
    }

    fn handle_event(&mut self, event: &Event, manager: &mut Manager) -> ScreenAction {
        if let Event::Key(KeyEvent { code, kind: KeyEventKind::Press, modifiers, .. }) = event {
            match code {
                KeyCode::Esc => return ScreenAction::Pop,
                KeyCode::Enter => {
                    // If nothing selected but filter has text, apply filter text directly
                    if self.list_state.selected().is_none() {
                        let typed = self.filter.trim().to_string();
                        if !typed.is_empty() {
                            return self.apply_branch(&typed, manager);
                        }
                    } else {
                        return self.apply_selected(manager);
                    }
                }
                KeyCode::Down | KeyCode::Char('j') => {
                    let next = match self.list_state.selected() {
                        Some(i) if i + 1 < self.display.len() => i + 1,
                        Some(i) => i,
                        None if !self.display.is_empty() => 0,
                        None => return ScreenAction::None,
                    };
                    self.list_state.select(Some(next));
                }
                KeyCode::Up | KeyCode::Char('k') => {
                    let prev = match self.list_state.selected() {
                        Some(0) | None => 0,
                        Some(i) => i - 1,
                    };
                    self.list_state.select(Some(prev));
                }
                _ => {
                    self.handle_filter_key(*code, *modifiers);
                    self.error.clear();
                }
            }
        }
        ScreenAction::None
    }

    fn is_modal(&self) -> bool {
        true
    }
}
